dex/src/main/java/io/crate/data/AsyncFlatMapper.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.data;
+
+import java.util.Iterator;
+import java.util.concurrent.CompletableFuture;
+
+public interface AsyncFlatMapper<I, O> extends AutoCloseable {
+
+    CompletableFuture<? extends Iterator<O>> apply(I item, boolean isLastCall);
```

Comment:
It is not so easy to see the `flatmap` here, seems like this only provides the map operation and the rest happens in `AsyncFlatMapBatchIterator`. However this makes all sense. Maybe a few lines of documentation/comments would be good.

Reply:
Yep, this is modelled after the stream API:

```
flatMap(Function<? super RewriteToQueryThenFetch,? extends Stream<? extends R>> mapper) : Stream<R>
```

The function that does the mapping is called `mapper`.

I will add some more docs to clarify that.

Reply:
I tried to add something, but it read as a repitition of the signature, so I'll keep it as it is for now.